src/core/geometry/qgsnurbscurve.cpp

```diff
@@ -0,0 +1,1622 @@
+      if ( i < mWeights.size() )
+        newWeights.append( mWeights[i] );
+    }
+  }
+
+  const bool changed = ( newPoints.size() != mControlPoints.size() );
+  mControlPoints = newPoints;
+  mWeights = newWeights;
+
+  return changed;
+}
+
+double QgsNurbsCurve::vertexAngle( QgsVertexId vertex ) const
+{
+  std::unique_ptr<QgsLineString> line( curveToLine() );
+  if ( !line )
+    return 0.0;
+  return line->vertexAngle( vertex );
+}
+
+void QgsNurbsCurve::swapXy()
+{
+  for ( QgsPoint &pt : mControlPoints )
+  {
+    const double x = pt.x();
+    pt.setX( pt.y() );
+    pt.setY( x );
+  }
+  clearCache();
+}
+
+bool QgsNurbsCurve::transform( QgsAbstractGeometryTransformer *transformer, QgsFeedback *feedback )
+{
+  Q_UNUSED( feedback );
+  if ( !transformer )
+    return false;
+
+  for ( QgsPoint &pt : mControlPoints )
+  {
+    double x = pt.x(), y = pt.y(), z = pt.z(), m = pt.m();
+    if ( !transformer->transformPoint( x, y, z, m ) )
+      return false;
+    pt.setX( x );
+    pt.setY( y );
+    pt.setZ( z );
+    pt.setM( m );
+  }
+
+  clearCache();
+  return true;
+}
+
+QgsAbstractGeometry *QgsNurbsCurve::createEmptyWithSameType() const
+{
+  return new QgsNurbsCurve();
+}
+
+double QgsNurbsCurve::closestSegment( const QgsPoint &pt, QgsPoint &segmentPt, QgsVertexId &vertexAfter, int *leftOf, double epsilon ) const
+{
+  std::unique_ptr<QgsLineString> line( curveToLine() );
+  if ( !line )
+  {
+    segmentPt = QgsPoint();
+    vertexAfter = QgsVertexId();
+    if ( leftOf )
+      *leftOf = 0;
+    return -1;
+  }
+  return line->closestSegment( pt, segmentPt, vertexAfter, leftOf, epsilon );
+}
+
+void QgsNurbsCurve::transform( const QgsCoordinateTransform &ct, Qgis::TransformDirection d, bool transformZ )
+{
+  for ( QgsPoint &pt : mControlPoints )
+  {
+    double x = pt.x();
+    double y = pt.y();
+    double z = transformZ && pt.is3D() ? pt.z() : std::numeric_limits<double>::quiet_NaN();
+    ct.transformInPlace( x, y, z, d );
+    pt.setX( x );
+    pt.setY( y );
+    if ( transformZ && pt.is3D() )
+      pt.setZ( z );
+  }
+  clearCache();
+}
+
+void QgsNurbsCurve::transform( const QTransform &t, double zTranslate, double zScale, double mTranslate, double mScale )
+{
+  for ( QgsPoint &pt : mControlPoints )
+  {
+    const QPointF p = t.map( QPointF( pt.x(), pt.y() ) );
+    pt.setX( p.x() );
+    pt.setY( p.y() );
+
+    if ( pt.is3D() )
+      pt.setZ( pt.z() * zScale + zTranslate );
+    if ( pt.isMeasure() )
+      pt.setM( pt.m() * mScale + mTranslate );
+  }
+  clearCache();
+}
+
+QgsRectangle QgsNurbsCurve::boundingBox() const
+{
+  return boundingBox3D().toRectangle();
+}
+
+QgsBox3D QgsNurbsCurve::boundingBox3D() const
+{
+  if ( mBoundingBox.isNull() )
+  {
+    mBoundingBox = calculateBoundingBox3D();
+  }
+  return mBoundingBox;
+}
+
+QgsBox3D QgsNurbsCurve::calculateBoundingBox3D() const
+{
+  if ( mControlPoints.isEmpty() )
+    return QgsBox3D();
+
+  const QgsPoint &first = mControlPoints[0];
```

Comment:
it seems to me that the envelope of a NURBS can overshoot the envelope of its control points ? I guess you need to curveToLine() first
(that said the same remark could be done on a CircularString and as far as I can see CircularString doesn't specialize boudingBox() and just rely on iterating over vertices, so with same issue)